lib/kernel/src/file.erl

```diff
@@ -397,25 +397,38 @@ write_file(Name, Bin) ->
       Modes :: [mode()],
       Reason :: posix() | badarg | terminated | system_limit.
 
-write_file(Name, Bin, ModeList) when is_list(ModeList) ->
-    case make_binary(Bin) of
-	B when is_binary(B) ->
-	    case open(Name, [binary, write | 
-			     lists:delete(binary, 
-					  lists:delete(write, ModeList))]) of
-		{ok, Handle} ->
-		    case write(Handle, B) of
-			ok ->
-			    close(Handle);
-			E1 ->
-			    _ = close(Handle),
-			    E1
-		    end;
-		E2 ->
-		    E2
-	    end;
-	E3 ->
-	    E3
+write_file(Name, IOData, ModeList) when is_list(ModeList) ->
+    case lists:member(raw, ModeList) of
+        true ->
+          %% For backwards compatibility of error messages
+            try iolist_size(IOData) of
+                _Size -> do_write_file(Name, IOData, ModeList)
+            catch
+                error:Error -> {error, Error}
+            end;
+        false ->
+            case make_binary(IOData) of
+                Bin when is_binary(Bin) ->
+                    do_write_file(Name, Bin, ModeList);
+                Error ->
+                    Error
+            end
+    end.
+
+do_write_file(Name, IOData, ModeList) ->
+    case open(Name, [binary, write |
```

Comment:
The calls to `lists:delete/2` are not necessary; duplicated options will be ignored.

I know that the original code does the same thing, but there is no reason to keep the duplication in your refactored code.